cli/cmd/upgrade.go

```diff
@@ -0,0 +1,207 @@
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/golang/protobuf/jsonpb"
+	pb "github.com/linkerd/linkerd2/controller/gen/config"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/linkerd/linkerd2/pkg/tls"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type (
+	upgradeOptions struct{ *installOptions }
+)
+
+func newUpgradeOptionsWithDefaults() *upgradeOptions {
+	return &upgradeOptions{newInstallOptionsWithDefaults()}
+}
+
+func newCmdUpgrade() *cobra.Command {
+	options := newUpgradeOptionsWithDefaults()
+
+	cmd := &cobra.Command{
+		Use:   "upgrade [flags]",
+		Short: "Output Kubernetes configs to upgrade an existing Linkerd control plane",
+		Long:  "Output Kubernetes configs to upgrade an existing Linkerd control plane.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			k, err := newKubernetes()
+			if err != nil {
+				return err
+			}
+
+			values, configs, err := options.buildFromCluster(k)
+			if err != nil {
+				return err
+			}
+
+			return values.render(os.Stdout, configs)
+		},
+	}
+
+	options.configure(cmd)
+	return cmd
+}
+
+func newKubernetes() (*kubernetes.Clientset, error) {
+	api, err := k8s.NewAPI(kubeconfigPath, kubeContext)
+	if err != nil {
+		return nil, err
+	}
+
+	return kubernetes.NewForConfig(api.Config)
+}
+
+// fetchInstallValuesFromCluster checks the kubernetes API to fetch an existing
+// linkerd configuration.
+//
+// This bypasses the public API so that we can access secrets and validate permissions.
```

Comment:
we should consider modifying `fetchConfigsOrDefault` to do this as well. i'd prefer we have a single code path for retrieving configs.

Reply:
The upgrade logic is different than the other uses.